File: dev-topics-codingexams/dev-topics-pintr-expand/PPINTR/repeatKeys.py
import collections

"""
Write a function that expands a HashMap to an array. The HashMap contains a key and a value, from which your function
will produce an array with the key repeated the respective number of times. Output order of the array does not matter.
 
Convert a dictionary with key:count pairs to an array where each array entry has the key "value" replicated count times.
"""


def _build_frequencies(text):
    freq = dict()
    if text is None:
        return freq
    txt = str(text).strip(" ")
    if len(txt) < 1:
        return freq
    words = text.split()
    if len(words) < 1:
        return freq
    return collections.Counter(words)


def _repeat_key(occurrence):
    repeats = []
    for key in occurrence:
        repeats.append(str(key) * occurrence[key])
    return repeats


def run_test(label, input_str):
    print("\n" + label + ":\n  - " + str(input_str))
    freq_by_key = _build_frequencies(input_str)
    print("  - Frequencies: " + str(freq_by_key))
    keys = _repeat_key(freq_by_key)
    print("  - Keys       : " + str(keys))


if __name__ == '__main__':
    run_test("None", None)
    run_test("Empty", "")
    run_test("One word", "Hello")
    run_test("Unique words", "Hello There!")
    doi = "We really really need to be really really faster than our " \
          "competition who need to be slower than us to win"
    run_test("Some Repeats", doi)

File: dev-topics-codingexams/dev-topics-pintr-expand/src/main/java/demo/pintr/expand/RepeatKeys.java
package demo.pintr.expand;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Write a function that expands a HashMap to an array. The HashMap contains a
 * key and a value, from which  your function will produce an array with the key
 * repeated the respective number of times. Output order of the array does not matter.
 *
 * @author Donald Trummell
 */
public class RepeatKeys {

	public static void main(String[] args) {
		runTest("Null", null);
		runTest("Empty", "");
		runTest("One word", "Hello");
		runTest("Unique words", "Hello There!");
		String doi = "We really really need to be really really faster than our "
				+ "competition who need to be slower than us for us to win";
		runTest("Some Repeats", doi);
	}

	private static List<String> repeatKey(Map<String, Integer> occurrence) {
		List<String> repeats = new ArrayList<String>();
		for (Map.Entry<String, Integer> e : occurrence.entrySet()) {
			repeats.add(Support.repeatString(e.getKey(), e.getValue()));
		}
		return repeats;
	}

	private static void runTest(String label, String input) {
		System.out.println("\n" + label + ":\n  - " + String.valueOf(input));
		Map<String, Integer> freqsByKey = Support.buildFrequencies(input);
		System.out.println("  - Frequencies: " + String.valueOf(freqsByKey));
		List<String> keys = repeatKey(freqsByKey);
		System.out.println("  - Keys       : " + String.valueOf(keys));
	}
}

class Support {
	static Map<String, Integer> buildFrequencies(String text) {
		Map<String, Integer> freq = new HashMap<String, Integer>();
		if (text == null || text.isEmpty()) {
			return freq;
		}

		String[] words = text.trim().split(" ");
		if (words.length < 1) {
			return freq;
		}

		for (int i = 0; i < words.length; i++) {
			freq.merge(words[i], 1, (x, y) -> x + y);
		}

		return freq;
	}

	// Java-11 has this function
	static String repeatString(String s, int n) {
		if (s == null || s.isEmpty() || n < 1) {
			return "";
		}

		StringBuilder sb = new StringBuilder(s);
		while (--n > 0) {
			sb.append(s);
		}

		return sb.toString();
	}
}
